### PR TITLE
Global linear bandit

### PR DESCRIPTION
**Pull request by choltz95**
Wanted to better familiarize myself with the code - please let me know if anything is off or if anything violates your goals for the lib.

I added support for a correlated/global linear bandit environment and C-UCB - this on/off set technique could be useful for heterogenous multiplayer bandits.

Had to include a cvxpy dependency in the learner & arm to solve line 5 and 6 from alg 1. in [1]  - not sure if this is okay with you. The solver makes the learner much slower compared to UCB, but the result w/ correlated arms is better.

Screen Shot 2020-02-13 at 12 20 59 PM

Currently running the full experiment to get a better time estimate since the slowdown may or may matter for scaled problems (~200 s for T=500, # trials=50 compared to <1 s for UCB on my macbook)

[0] https://arxiv.org/abs/1911.03959
[1] https://users.ece.cmu.edu/~oyagan/Conferences/gupta2018structured.pdf

### CODE REVIEW

File: arms/lineararm.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
+import numpy as np
+import cvxpy as cp
+from .utils import Arm
+
+__all__ = ['LinearArm']
+
+
+class LinearArm(Arm):
```
Comment (Alanthink) [inline arms/lineararm.py]: Is it better to name it "CorrelatedArm"? It is confusing considering the well-known linear bandit.

File: bandits/ordinarylinearbandit.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+  def arm_num(self):
+    """return number of arms"""
+    return self.__arm_num
+
+  @property
+  def arms(self):
+    return self.__arms
+
+  @property
+  def type(self):
```
Comment (Alanthink) [inline bandits/ordinarylinearbandit.py]: Here should return the name of the bandit.

File: bandits/ordinarylinearbandit.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+        for tup in enumerate(self.__arms)], key=lambda x:x[1])[0]
+    self.__best_arm = self.__arms[self.__best_arm_ind]
+
+  @property
+  def arm_num(self):
+    """return number of arms"""
+    return self.__arm_num
+
+  @property
+  def arms(self):
```
Comment (Alanthink) [inline bandits/ordinarylinearbandit.py]: Should this property be protected?

File: examples/linear_bandits_config.json [hunk 4]
```diff
@@ -0,0 +1,23 @@
+{
```
Comment (Alanthink) [inline examples/linear_bandits_config.json]: I am trying to change the config style such that a bandit will accept all environment parameters. In this way, we do not need to write a new program to parse the environment parameters. Feel free to change the `main.py` file. I will take care of the left changes associated with other bandit settings.

```
"environment": {
	"type": "CorrelatedBandit",
	"CorrelatedBandit": {
		"arm": {
			"type": "CorrelatedArm",
			"means": [[0.3,0.2], [0.5,0.6], [0.7,0.9]]
		},
		"param": [0.5,0.5]
	}
}
```

Comment (Alanthink) [reply examples/linear_bandits_config.json]: I have already updated the `main.py` file such that arm setup in now in bandit initialization.

File: bandits/ordinarylinearbandit.py (hunk 3 above)
Comment (choltz95) [reply bandits/ordinarylinearbandit.py]: So one question that I had: for this correlated/linear bandits, each action corresponds to a vector (defined in the config under 'means' key) that is known to the player.  I unprotected this arm property so I could initialize this vector for em_arms in cucb.py line 27.
```
for k in range(self._arm_num):
      self._em_arms[k].action = self._bandit.arms[k].action
```
What would be the correct way that you recommend to pass this action vector that is property of the arm to the empirical arm so the learner can reference it?

Comment (Alanthink) [reply bandits/ordinarylinearbandit.py]: At this point, I suggest putting a property under `bandit` class and let the learner communicate only with the bandit. Also, can you pull my new commit? In my new commit, I update the config file to the following style and arm setup is now put under the bandit initialization. And then you can define a property telling the learners that what actions are.

```
{
	"environment": {
                "type": "OrdinaryCorrelatedBandit",
		"OrdinaryCorrelatedBandit": {	
			"type": "CorrelatedArm",
			"means": [[0.3,0.2], [0.5,0.6], [0.7,0.9]],
                        "param": [0.5,0.5]
		}
	},
	"learner": {
		"goal": "regretmin",
		"type": "ordinarylearner",
		"policy": ["Uniform","CUCB","UCB"]
	},
	"parameters": {
		"horizon": 2000,
		"freq": 20,
		"trials": 200,
		"processors": 40
	}
}
```

Comment (choltz95) [reply bandits/ordinarylinearbandit.py]: Commit a refactor for compatibility w/ the config file & added actions property under CorrelatedBandit class, separated correlatedlearners.

File: bandits/correlatedbandit.py [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+
+  def init(self):
+    self.__tot_samples = 0
+    self.__max_rewards = 0
+
+  @property
+  def context(self):
+    return None
+
+  @property
```
Comment (Alanthink) [inline bandits/correlatedbandit.py]: This var should not be accessible outside the bandit

File: bandits/correlatedbandit.py [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+        for tup in enumerate(self.__arms)], key=lambda x:x[1])[0]
+    self.__best_arm = self.__arms[self.__best_arm_ind]
+
+  @property
+  def arm_num(self):
+    """return number of arms"""
+    return self.__arm_num
+
+  @property
+  def arms(self):
```
Comment (Alanthink) [inline bandits/correlatedbandit.py]: Not useful anymore.

File: bandits/correlatedbandit.py [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+    logging.info('Correlated bandit model')
+    if pars['arm']['type'] !=  'CorrelatedArm':
+      logging.fatal('Not a correlated arm!')
+    actions = pars['arm']['means']
+    if not isinstance(actions, list):
+      logging.fatal('Means should be given in a list!')
+    self.__actions = [np.array(action) for action in actions]
+    self.__theta = np.array(pars['param'])
+    Arm = getattr(import_module(ARM_PKG), pars['arm']['type'])
+    arms = [Arm(np.dot(action,self.theta)) for action in self.__actions]
```
Comment (Alanthink) [inline bandits/correlatedbandit.py]: `self.theta` -> `self.__theta`

File: bandits/correlatedbandit.py [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+    self.__arms = arms
+
+    for idx, action in enumerate(self.__actions):
+      if action.shape !=self.theta.shape:
```
Comment (Alanthink) [inline bandits/correlatedbandit.py]: Ditto

File: bandits/correlatedbandit.py [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+    self.__tot_samples = 0
+    self.__max_rewards = 0
+
+  @property
+  def context(self):
+    return None
+
+  @property
+  def theta(self):
+    return self.__theta
```
Comment (Alanthink) [inline bandits/correlatedbandit.py]: This quantity should not be exposed.

File: learners/regretmin/correlatedlearner/cucb.py [hunk 10]
```diff
@@ -0,0 +1,64 @@
+import numpy as np
+import cvxpy as cp
+
+from .utils import CorrelatedLearner
+
+__all__ = ['CUCB']
+
+
+class CUCB(CorrelatedLearner):
+  """UCB"""
```
Comment (Alanthink) [inline learners/regretmin/correlatedlearner/cucb.py]: `CUCB`

**Pull request merged**